### PR TITLE
Move BOARD_* and SOC_* configuration to unselectable Kconfigs

### PR DESCRIPTION
**Pull request by nordicjm**
These commits change how BOARD_* and SOC_* Kconfig variables are selected from being freely selectable in a build, e.g. if a build is configured for an STM32 platform then from menuconfig, it is possible to change to an atmel platform which results in a completely broken build - to a system whereby those variables can no longer be changed after a project is configured.

This is needed for future sysbuild integration work whereby filtering can be performed from a SOC_* and ARCH_* level.

Note: Conversion of nordic soc is currently absent

### CODE REVIEW

File: soc/arm/nuvoton_npcx/npcx7/Kconfig.soc [hunk 1]
```diff
@@ -3,17 +3,17 @@
...
+	help
+	  NPCX7M6FC
 
 config SOC_NPCX7M7FC
-	bool "NPCX7M7FC"
-
-endchoice
+	bool
+	help
+	  NPCX7M7FC
```
Comment (tejlmand) [inline soc/arm/nuvoton_npcx/npcx7/Kconfig.soc:19]: i'm very much in favor of `help` texts, especially because they are rendered into the official documentation, but a help text which simply just adds the SoC name is not of much value, so for those I prefer to simply keep the config as-is, that is without a help text.

File: Kconfig.zephyr [hunk 2]
```diff
@@ -19,7 +19,7 @@ osource "$(KCONFIG_BINARY_DIR)/Kconfig.shield.defconfig"
 # This loads Zephyr base shield defconfigs
 source "boards/shields/*/Kconfig.defconfig"
 
-source "$(BOARD_DIR)/Kconfig.defconfig"
+osource "$(BOARD_DIR)/Kconfig.defconfig"
```
Comment (tejlmand) [inline Kconfig.zephyr:22]: this effectively changes `Kconfig.defconfig` from being mandatory to being optional.

The alternative approach is to keep the `source` and then boards not having any default to override  must create an empty file, maybe containing `# Intentionally left empty`.

I have a slight preference for keeping the `source` and then an empty file for those boards not needing to override anything.

Reason being:
- most downstream users of Zephyr will at some point need to create a custom board of there own, and having a `source` helps them to place board specific defconfigs the right place.
- It prevents stupid user mistakes like writing `kconfig.defconfig` / `Kconfig.defconf`instead of `Kconfig.defconfig`.
- If board is later updated to override defaults then it's clearer where to do this if the file is already empty (but present)

**Review by tejlmand — COMMENTED** [review]
quick early glance.

Minor nit on the commit messages:
> Make SOC Kconfig options unselectable

technically speaking the setting is still selectable (using `select <config>`), but it's not user selectable.

So a small proposal:
> Make SOC Kconfig options internal

**Comment by nordicjm** [comment]
@stephanosio Is it possible to get an exclusion on needing copyright headers on these Kconfig files?

**Comment by stephanosio** [comment]
> @stephanosio Is it possible to get an exclusion on needing copyright headers on these Kconfig files?

To be discussed here: https://discord.com/channels/720317445772017664/989195595572973648/1019570877895036970

**Comment by stephanosio** [comment]
> Move BOARD_* and SOC_* configuration to unselectable Kconfigs

**promptless** would be a more correct wording.

File: soc/arm/nuvoton_npcx/npcx7/Kconfig.soc (hunk 1 above)
Comment (nashif) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.soc:19]: that does not apply everywhere, `Nuvoton M48X Series MCU` is much more descriptibe than `SOC_SERIES_M48X`. This  commit is removing valuable information IMO.

Comment (tejlmand) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.soc:19]: i'm only requesting not to add dummy help text where it is not existing.
(or improve the help text to be meaningful)

SoCs that has relevant help text should definitely keep that.

Comment (nordicjm) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.soc:19]: The text was already there, in the selection field, this just moves it from the selection field (as it is no longer selectable) to the help field

File: Kconfig.zephyr (hunk 2 above)
Comment (tejlmand) [reply Kconfig.zephyr:22]: fair enough, we have different opinion on this.

Another risk is that someone later wants to add a new `default` to a board which doesn't have a `Kconfig.defconfig` already.
Contributor looks around in that board folder and decides to place it in `Kconfig.$(BOARD)` because lack of knowledge.
Such PR can easily approved and be merged, resulting in inconsistency.

Whereas a `Kconfig.defconfig` with content:
```
# Kconfig.defconfig file for board specific overload of default settings.

# No board specific defaults.
```

will help to ensure that such defaults goes to the right location.

Contributors will often look for an existing file to place small suggestions, rather than create new files.

**Review by tejlmand — COMMENTED** [review]
if in anyway possible, we should only deprecated the old way of handling boards (and SoCs), as to still allow downstream users to keep using their existing, and old style boards.

A deprecation warning could be issued if we discover old style `Kconfig.board` and a missing `Kconfig.${BOARD}` file.

Feel free to ping directly for ideas on how this might be achieved, or provide reasons if this is not possible.

**Comment by nordicjm** [comment]
> if in anyway possible, we should only deprecated the old way of handling boards (and SoCs), as to still allow downstream users to keep using their existing, and old style boards.
> 
> A deprecation warning could be issued if we discover old style `Kconfig.board` and a missing `Kconfig.${BOARD}` file.
> 
> Feel free to ping directly for ideas on how this might be achieved, or provide reasons if this is not possible.

This is not supported, boards must be updated, as old boards will be using *_defconfig files to select SOC symbols which is no longer possible, they need to move selections to Kconfig* files

**Review by andyross — CHANGES_REQUESTED** [review]
Jumping into the bikeshed for a handful of comments, though I won't pursue.  I don't understand the point here.  I think I see what the PR is doing, but the justification seems pretty weak for me: "it is possible to [change platforms in menuconfig] which results in a completely broken build"

... well, yeah?  I mean, you can completely break your build in any number of ways with kconfig (and dts, and cmake, and now sysbuild, and frankly by messing with C code too!) in ways that this PR doesn't try to address.  Are we really going to come in with a shotgun to do 2040-file refactorings to prevent every possible mistake?

Obviously we aren't.  So what's the real reason?  There's something about filtering for sysbuild in the comment.   Can you detail that and why it requires this?

Also, this scheme seems kinda fragile to me.  You want the platform kconfigs to conform to specific rules, but you're not adding any framework to enforce that or to detect violations.  There's absolutely no way this is going to stick, new lpatform authors (and old curmudgeons like me) are never going to understand these rules and you'll be chasing fixups for years.

I think this rises to a -1 from my perspective based on the size of the PR.  But I don't think it's worthwhile inserting myself into a giant argument, so feel free to break the review if there's consensus in the opposite direction.

As an alternative: maybe what you really want is to move "platform" out of kconfig and have it entirely managed by the cmake layer instead?  Then cmake could auto-generate the needed kconfig settings for arch/soc/series/board in a way that conforms to the scheme you want?  That would likewise be a giant patch, but it would be almost all code **removal** (i.e. stuff Andy loves) and wouldn't have the mistake-prone problem of having to enforce a complicated design scheme across thousands of files using a language (kconfig) that doesn't really have support for that kind of framework.

**Comment by nordicjm** [comment]
There are multiple interconnected issues that this PR serves to address:

1. The CMake board variable name does not match the Kconfig board name, this can be see most easily when building hello_world for an ARM Cortex M33 board running on the _ns part of it, the BOARD name will be lacking the _ns part - this is a bug
2. SOC selection is a feature which can be added to sysbuild, at present, sysbuild has no knowledge of the SOC it is being built for, this means if you have a configuration with multiple cores, or a core requiring an external library e.g. a pre-compiled bootloader (which may be any bootloader) or security library, to add support for managing this would need to be done on a board-by-board basis at present, but in most cases it is not a feature of a board, it is a feature of a SOC. To allow sysbuild to perform filtering for SOC configuration, this means that the SOC_* Kconfig values need to be loaded in the main sysbuild system so that Kconfig and CMake configuration can be performed, the option can be added to load multiple bootloaders for different cores, or to load an external image for example a sound DSP library which is external to a project. In order to do this, the SOC_* Kconfig variables need to be present in the sysbuild menu without things like configuring the CPU options, because those are done on a per-project basis, this necessitated removal of choice options (which, coincidentally, happens to resolve issue 3 below also) so that the options could be present in the build, to get this working, the SOC_* Kconfig options needs to be moved from the boards *_defconfig files to their Kconfig files because the *_defconfig files contain a lot of selections that are not present in sysbuild, for example to enable a low frequency clock source, or to enable a UART - these are not present and have no use in it. With this addition, sysbuld can be expanded to filter and include projects (including external, non-zephyr projects) for SOCs.
3. Menuconfig allows changing of the processor type which breaks that build because it mixes unsupported Kconfig values from the previous selected processor with the current, this might seem like a minor inconvenience but why is the option present to change selectable at all? You might not use it because you know it will break, but newcomers to the project don't know what, I certainly didn't know that years ago when I used it and had a project which I was manually configuring suddenly need to be cleared and reconfigured again. This bug is resolved by keeping the Kconfigs present but not allowing them to be changed, so CMake/Kconfig code that makes decisions based off selections will still work with this as it does today

> Also, this scheme seems kinda fragile to me. You want the platform kconfigs to conform to specific rules, but you're not adding any framework to enforce that or to detect violations. There's absolutely no way this is going to stick, new lpatform authors (and old curmudgeons like me) are never going to understand these rules and you'll be chasing fixups for years.

What new rules? The SOC selections go into Kconfig files instead of the _defconfig files, this will be immediately obvious if anyone gets it wrong because a warning will be emitted that an option is present in the _defconfig file which is sourced from other values and not directly settable, 

> As an alternative: maybe what you really want is to move "platform" out of kconfig and have it entirely managed by the cmake layer instead? Then cmake could auto-generate the needed kconfig settings for arch/soc/series/board in a way that conforms to the scheme you want? That would likewise be a giant patch, but it would be almost all code removal (i.e. stuff Andy loves) and wouldn't have the mistake-prone problem of having to enforce a complicated design scheme across thousands of files using a language (kconfig) that doesn't really have support for that kind of framework.

The method zephyr is using is no different than the method that the linux kernel itself is using, I do not agree that replacing all the Kconfig parts with CMake parts would be the right solution for this or add anything of value.

**Comment by andyross** [comment]
Some drive-by comments to clarify my position, though I really don't want to be dragged into a bikeshed and promise I'll resist continuing:

> SOC selection is a feature which can be added to sysbuild, at present, sysbuild has no knowledge of the SOC it is being built for

So... platform selection might need to be at a higher level than raw kconfig, right?  If it was something you could get via cmake via a short-circuit/trivial target before the build starts, that would fix all your problems, right?

> The method zephyr is using is no different than the method that the linux kernel itself is using

But it is!  Linux specifies architecture via an ARCH= variable passed to make[1], not via kconfig (though it mirrors it in kconfig).  And while I don't know the history I strongly suspect it's for exactly this reason, to fix the chicken-and-egg dependency issues between platform layers that can't port like that.  And this is exactly what I'm suggesting you investigate if you have the same problem.

Basically: if we're looking at a 2000-file PR, we need to be really sure we're solving the problem in the best way.  I don't think this is what we want.

[1] And to be fair, Linux doesn't even try to solve the whole problem.  The Linux equivalent to sysbuild is a bunch of distro/OS-supplied kernel/initrd packing/signing tools.  And those know their platform a priori, they don't demand that Linux tell them.  Which... is something I guess I'm not understanding about sysbuild, shouldn't it be **starting** with a solution to this problem rather than trying to squeeze it out of Zephyr?

**Comment by nordicjm** [comment]
> But it is! Linux specifies architecture via an ARCH= variable passed to make[1], not via kconfig (though it mirrors it in kconfig). And while I don't know the history I strongly suspect it's for exactly this reason, to fix the chicken-and-egg dependency issues between platform layers that can't port like that. And this is exactly what I'm suggesting you investigate if you have the same problem.

We do the same but from a board concept instead of arch, `-DBOARD=x`, hence why with the introduction of this PR you can no longer change the board, because doing so breaks the build, likewise in linux to change the arch would break the build, so needs to be done afresh.
